gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java

```diff
@@ -0,0 +1,185 @@
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.Sets;
+import com.google.common.collect.Streams;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.maven.shared.dependency.analyzer.ClassAnalyzer;
+import org.apache.maven.shared.dependency.analyzer.DefaultClassAnalyzer;
+import org.apache.maven.shared.dependency.analyzer.DependencyAnalyzer;
+import org.apache.maven.shared.dependency.analyzer.asm.ASMDependencyAnalyzer;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.GradleException;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ResolvedArtifact;
+import org.gradle.api.artifacts.ResolvedDependency;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.TaskAction;
+
+@SuppressWarnings("DesignForExtension")
+public class CheckExactDependenciesTask extends DefaultTask {
+
+    private static final ClassAnalyzer classAnalayzer = new DefaultClassAnalyzer();
+    private static final DependencyAnalyzer dependencyAnalyzer = new ASMDependencyAnalyzer();
+
+    private List<Configuration> dependenciesConfigurations = new ArrayList<>();
+    private FileCollection classes;
+
+    private Map<String, ResolvedArtifact> classToDependency = new HashMap<>();
+    private Map<ResolvedArtifact, Set<String>> classesFromArtifact = new HashMap<>();
+    private Map<ResolvedArtifact, ResolvedDependency> artifactsFromDependency = new HashMap<>();
+
+    public CheckExactDependenciesTask() {
+        setGroup("Verification");
+    }
+
+    @TaskAction
+    public void checkUnusedDependencies() {
+        Set<ResolvedDependency> declaredDependencies = getDependenciesConfigurations().stream()
+                .map(Configuration::getResolvedConfiguration)
+                .flatMap(resolved -> resolved.getFirstLevelModuleDependencies().stream())
+                .collect(Collectors.toSet());
+
+        // All artifacts
+        Set<ResolvedArtifact> allArtifacts = declaredDependencies.stream()
+                .flatMap(dependency -> dependency.getAllModuleArtifacts().stream())
+                .collect(Collectors.toSet());
+
+        // Construct caches of classes in each dependency
+        allArtifacts.forEach(artifact -> {
+            try {
+                // Construct class/artifact maps
+                Set<String> classesInArtifact = classAnalayzer.analyze(artifact.getFile().toURI().toURL());
+                classesFromArtifact.put(artifact, classesInArtifact);
+                classesInArtifact.forEach(clazz -> {
+                    if (classToDependency.put(clazz, artifact) != null) {
+                        getLogger().info("Found duplicate class " + clazz);
```

Comment:
`getLogger().info("Found duplicate class {}", clazz);`

Reply:
might be worth including the artifacts with duplicate classes as well